### PR TITLE
CLN/STYLE: Lint comprehensions

### LINKED ISSUE #18464: STYLE: Lint for comprehensions that use the list/set/dict functions
xref #18383

Ideally comprehensions should follow the general pattern:

`list(x for x in iterator)` --> `[x for x in iterator]`
`set(x for x in iterator)` --> `{x for x in iterator}`
`dict((x,x) for x in iterator)` --> `{x: x for x in iterator}`

Adding a lint check for the prior pattern in `ci/lint.sh` would be great.

### PR DESCRIPTION
**Pull request by mroeschke**
- [x] closes #18464
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

Using flake8-comprehensions (xref https://github.com/pandas-dev/pandas/issues/20588#issuecomment-405057018) to lint for unnecessary comprehensions. I just added this to the 2.7-travis build since it appears to be our linting build.

### CODE REVIEW

**Comment by jbrockmendel** [comment]
Syntax error in one of the edits.  +1 on the concept.

File: ci/travis-27.yaml [hunk 1]
```diff
@@ -9,6 +9,7 @@ dependencies:
   - fastparquet
   - feather-format
   - flake8=3.4.1
+  - flake8-comprehensions
```
Comment (jreback) [inline ci/travis-27.yaml]: need to add this to the dev deps as well

**Comment by jbrockmendel** [comment]
Tons of flake complaints, may need to implement gradually.

**Comment by jbrockmendel** [comment]
Now it’s mostly upset about lambdas.  OTOH it’s reassuring to see mine isn’t the only PR that pytables is messing up.

**Comment by mroeschke** [comment]
Is there a predefined set of lint codes we ignore?

I am a little puzzled why these lint checks now started to fire; I figure it has something to do with the scope of checks when `--ignore` is specified.

**Comment by jbrockmendel** [comment]
> Is there a predefined set of lint codes we ignore?

in setup.cfg

**Comment by jbrockmendel** [comment]
Possibly for the while-were-at-it pile: 

#11954 checks for unnecessary backslashes (could this be done with just a grep?)
Enable checks for bare “except:” statements in a subset of files and gradually expand?

**Comment by jreback** [comment]
we have a fixed version of flake8 (an open issue about this too) - which doesn’t check bare excepts

**Comment by jreback** [comment]
pls rebase

**Comment by mroeschke** [comment]
The travis failure is from the reoccuring `ResourceWarning` that appears in `test_dialect_conflict`

**Comment by jreback** [comment]
thanks @mroeschke 

if you think need to open additional issues for the skipped pep codes, pls do. alternatively maybe just add some docs on them on why we need to keep them there (e.g. so next person isn't wondering why those particular codes).

File: ci/lint.sh [hunk 2]
```diff
@@ -10,42 +10,42 @@ if [ "$LINT" ]; then
...
-    flake8 pandas --filename=*.py --exclude pandas/_libs/src
+    flake8 pandas --filename=*.py --exclude pandas/_libs/src --ignore=C405,C406,C408,C409,C410,E402,E731,E741,W503
     if [ $? -ne "0" ]; then
         RET=1
     fi
     echo "Linting *.py DONE"
 
     echo "Linting setup.py"
-    flake8 setup.py
+    flake8 setup.py --ignore=C405,C406,C408,C409,C410,E402,E731,E741,W503
```
Comment (jreback) [inline ci/lint.sh]: e.g. here if you'd document why these codes (general comments are enough), though if specific ones *can* be removed pls do that. also *maybe* want to add these to ``setup.cfg`` ?

File: ci/environment-dev.yaml [hunk 3]
```diff
@@ -6,6 +6,7 @@ dependencies:
   - Cython>=0.28.2
   - NumPy
   - flake8
+  - flake8-comprehensions
```
Comment (jreback) [inline ci/environment-dev.yaml]: I think need to run and commit scripts/convert_deps.py to have requirements_dev.txt updated. can you PR that?

**Pull request merged**